centrallix-sysdoc/OSDriver_Authoring.md

```diff
@@ -115,497 +170,1032 @@ License:  Copyright (C) 2001-2011 LightSys Technology Services. See LICENSE.txt
 
-The SetAttrValue function works much the same way as GetAttrValue, just with the information moving in the opposite direction.  The third parameter, void* value, is treated in the same manner.
+| Parameter | Type          | Description
+| --------- | ------------- | ------------
+| inf_v     | void*         | A driver instance pointer (returned from `Open()` or `QueryFetch()`).
+| attr_name | char*         | The name of the requested attribute.
+| oxt       | pObjTrxTree*  | The transaction tree pointer for the `OBJDRV_C_TRANS` capability.
+
+The returns a new pObjPresentationHints struct on success, or NULL to indicate an error, in which case `mssError()` should be called before returning.  This struct should be allocated using `nmMalloc()`, and memset to zero, like this:
+```c
+pObjPresentationHints hints = nmMalloc(sizeof(ObjPresentationHints));
+if (hints == NULL) goto error_handling;
+memset(hints, 0, sizeof(ObjPresentationHints));
+```
 
-The GetFirstAttr and GetNextAttr functions each take two parameters, the inf_v pointer and the oxt* pointer, and are used to iterate through the non-mandatory attributes for the object.  GetFirstAttr should return a string naming the first attribute, and GetNextAttr should iterate through subsequent attributes.  When the attributes are exhausted, these functions should return NULL.  The attributes 'name', 'annotation', and 'content_type' should not be returned.  If the object has no other attributes, GetFirstAttr should return NULL.
+The return value, `hints : ObjPresentationHints`, contains the following useful fields which the function should set to give various useful information about the attribute.
+- `hints->Constraint : void*`: An expression for determining if a value is valid.
+- `hints->DefaultExpr : void*`: An expression defining the default value.
+- `hints->MinValue : void*`: An expression defining the minimum valid value.
+- `hints->MaxValue : void*`: An expression defining the maximum valid value.
+- `hints->EnumList : XArray`: If the attribute is a string enum, this XArray lists the valid string values.
+- `hints->EnumQuery : char*`: A query string which enumerates the valid values a string enum attribute.
+- `hints->Format : char*`: presentation format - datetime or money  <!-- TODO: Greg - We should add detail here. -->
+- `hints->AllowChars : char*`: An array of all valid characters for a string attribute, NULL to allow all characters.
+- `hints->BadChars : char*`: An array of all invalid characters for a string attribute.
+- `hints->Length : int`: The maximum length of data that can be included in a string attribute.
+- `hints->VisualLength : int`: The length that the attribute should be displayed if it is show to the user.
+- `hints->VisualLength2 : int`: The number of lines to use in a multi-line edit box for the attribute.
+- `hints->BitmaskRO : unsigned int`: which bits, if any, in bitmask are read-only <!-- TODO: Greg - Clarify which bitmask? -->
```

Comment:
Bitmask - this is for integer values that are bitmasks, for example the 'permissions' attribute (mode) of a file or a directory.

Reply:
That's entirely data-specific.  The BitmaskRO just specifies which ones cannot be modified.

Reply:
Ok, updated.